File: effective-kata-jvm/kata-wallet/src/main/java/com/effective/kata/wallet/Wallet.java
package com.effective.kata.wallet;

import java.math.BigDecimal;
import java.util.Currency;
import java.util.HashMap;
import java.util.Map;

public class Wallet {
    private final RateProvider rateProvider;
    private final Map<StockType, BigDecimal> stocks;

    public Wallet(RateProvider rateProvider) {
        this.rateProvider = rateProvider;
        this.stocks = new HashMap<>();
    }

    public BigDecimal value(Currency currency) {
        return stocks.entrySet().stream()
                .map(entry -> {
                    StockType stockType = entry.getKey();
                    BigDecimal quantity = entry.getValue();
                    BigDecimal rate = rateProvider.getRate(stockType, currency);
                    return quantity.multiply(rate);
                })
                .reduce(BigDecimal.ZERO, BigDecimal::add);
    }

    public void addStock(Stock stock) {
        BigDecimal currentQuantity = stocks.getOrDefault(stock.type(), BigDecimal.ZERO);
        stocks.put(stock.type(), currentQuantity.add(stock.quantity()));
    }
}

File: effective-kata-jvm/kata-wallet-kotlin/src/main/kotlin/com/effective/kata/wallet/Wallet.kt
package com.effective.kata.wallet

import java.math.BigDecimal
import java.math.BigDecimal.ZERO
import java.util.Currency

class Wallet {
    private val stocks: MutableMap<StockType, BigDecimal> = HashMap()

    fun addStock(stock: Stock): Wallet {
        val currentQuantity = stocks[stock.type] ?: ZERO
        val newQuantity = currentQuantity.add(stock.quantity)
        stocks[stock.type] = newQuantity
        return this
    }

    fun value(currency: Currency, rateProvider: RateProvider): BigDecimal {
        return stocks.map { (key, value) ->
            val price = rateProvider.getRate(key, currency)
            price.multiply(value)
        }.reduce(BigDecimal::add) ?: ZERO
    }
}